snorkel/mtl/model.py

```diff
@@ -0,0 +1,402 @@
+import logging
+import os
+from collections import defaultdict
+from collections.abc import Iterable
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+from snorkel.mtl.task import Task
+from snorkel.mtl.utils import move_to_device, prob_to_pred, recursive_merge_dicts
+
+model_default_config = {
+    "model_path": None,  # the path to a saved checkpoint to initialize with
+    "device": 0,  # gpu id (int) or -1 for cpu
+    "dataparallel": True,
+}
+
+
+class MultitaskModel(nn.Module):
+    """A class to build multi-task model.
+
+    :param name: Name of the model
+    :type name: str
+    :param tasks: a list of Task that trains jointly
+    :type tasks: list of Task project
+    """
+
+    def __init__(self, name=None, tasks=None, **kwargs):
+        super().__init__()
+        self.config = recursive_merge_dicts(
+            model_default_config, kwargs, misses="insert"
+        )
+        self.name = name if name is not None else type(self).__name__
+
+        # Initiate the model attributes
+        self.module_pool = nn.ModuleDict()
+        self.task_names = set()
+        self.task_flows = dict()
+        self.loss_funcs = dict()
+        self.output_funcs = dict()
+        self.scorers = dict()
+
+        # Build network with given tasks
+        if tasks is not None:
+            self._build_network(tasks)
+
+        logging.info(
+            f"Created multi-task model {self.name} that contains "
+            f"task {self.task_names}."
+        )
+
+        # Move model to specified device
+        self._move_to_device()
+
+    def _move_to_device(self):
+        """Move model to specified device."""
+
+        if self.config["device"] >= 0:
+            if torch.cuda.is_available():
+                logging.info(f"Moving model to GPU " f"(cuda:{self.config['device']}).")
+                self.to(torch.device(f"cuda:{self.config['device']}"))
+            else:
+                logging.info("No cuda device available. Switch to cpu instead.")
+
+    def _build_network(self, tasks):
+        """Build the MTL network using all tasks"""
+
+        if not isinstance(tasks, Iterable):
+            tasks = [tasks]
+        for task in tasks:
+            if task.name in self.task_names:
+                raise ValueError(
+                    f"Found duplicate task {task.name}, different task should use "
+                    f"different task name."
+                )
+            if not isinstance(task, Task):
+                raise ValueError(f"Unrecognized task type {task}.")
+            self.add_task(task)
+
+    def add_task(self, task):
+        """Add a single task into MTL network"""
+
+        # Combine module_pool from all tasks
+        for key in task.module_pool.keys():
+            if key in self.module_pool.keys():
+                if self.config["dataparallel"]:
```

Comment:
Should already be wrapped with `nn.DataParallel` given the below lines, right?  This is redundant?